google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/DependencyVersionWithBomInspection.java

```diff
@@ -0,0 +1,172 @@
+import com.intellij.util.xml.highlighting.DomElementsInspection;
+import java.util.Set;
+import org.jetbrains.annotations.Nls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.idea.maven.dom.model.MavenDomDependency;
+import org.jetbrains.idea.maven.dom.model.MavenDomProjectModel;
+
+/**
+ * An {@link DomElementsInspection} that detects google-cloud-java dependencies that have an
+ * explicit version definition when a BOM is defined.
+ *
+ * <p>Provides a quick-fix to strip out the version tag from the dependency.
+ */
+public class DependencyVersionWithBomInspection
+    extends DomElementsInspection<MavenDomProjectModel> {
+
+  private static final Logger logger = Logger.getInstance(AddCloudLibrariesAction.class);
+
+  public DependencyVersionWithBomInspection() {
+    super(MavenDomProjectModel.class);
+  }
+
+  @Nullable
+  @Override
+  public String getStaticDescription() {
+    return GctBundle.getString("cloud.libraries.version.with.bom.inspection.description");
+  }
+
+  @Override
+  public void checkFileElement(
+      DomFileElement<MavenDomProjectModel> domFileElement, DomElementAnnotationHolder holder) {
+    MavenDomProjectModel projectModel = domFileElement.getRootElement();
+
+    checkDependencyVersionWithBom(projectModel, holder);
+  }
+
+  /**
+   * Locates google-cloud-java dependencies that have a version tag defined when a BOM is also
+   * imported in the pom.xml. Displays an inspection warning and suggests a quickfix to delete the
+   * version from the dependency.
+   */
+  private void checkDependencyVersionWithBom(
+      MavenDomProjectModel projectModel, DomElementAnnotationHolder holder) {
+    Module module = projectModel.getModule();
+
+    if (module == null) {
+      return;
+    }
+
+    Set<CloudLibrary> cloudLibraries =
+        CloudLibraryProjectState.getInstance(module.getProject()).getCloudLibraries(module);
+
+    if (cloudLibraries.isEmpty()) {
+      return;
+    }
+
+    if (!CloudLibraryProjectState.getInstance(module.getProject())
+        .getCloudLibraryBom(module)
+        .isPresent()) {
+      return;
+    }
+
+    projectModel
+        .getDependencies()
+        .getDependencies()
+        .forEach(
+            dependency -> {
+              if (hasVersion(dependency) && isCloudLibraryDependency(dependency, cloudLibraries)) {
+                holder.createProblem(
+                    dependency.getVersion(),
+                    HighlightSeverity.GENERIC_SERVER_ERROR_OR_WARNING,
+                    GctBundle.message(
+                        "cloud.libraries.version.with.bom.inspection.problem.description"),
+                    new StripDependencyVersionQuickFix());
+              }
+            });
+  }
+
+  /** Checks to see if the {@link MavenDomDependency} has a version defined. */
+  private boolean hasVersion(MavenDomDependency dependency) {
+    return dependency.getVersion().exists();
```

Comment:
So if version element exists, this is definitely a warning - it cannot be empty, correct?

Reply:
are you asking if the version tag has an empty value, will it return true/false? Ill check this.

Reply:
exactly, thanks

Reply:
in the light of recent NPEs from using IJ APIs, can `dependency.getVersion()` be null in some case as well? :)

Reply:
there are only 2 cases I can think of:
1) the version tag is entirely missing
2) the version tag exists, but the value is empty

I know that case 1 is covered. I'll check on case 2 which was your original question, but I doubt it would trigger an NPE given case 1. I guess it still doesn't hurt to wrap it in a null check to be safe.